### PR TITLE
WIP: String Tensors

### PR DESCRIPTION
**Pull request by slyalin**
This is a "cumulative" PR for changes related to string tensor support. It won't be merged as-is, its purpose is to be a reference for other PRs that implement the things in a right way. Based on working PoC conducted 6 months ago.

It has some stuff that have already been brought as different PRs (like tokenizers). It requires clean up.

@rkazants: please look at core changes excluding infer request and new ops.
@dmitry-gorokhov: CPU part is for you
@jiwaszki: we will talk about Python part, just FYI that we are on the way; I need more preparation, please don't look inside at this moment.
@apaniukov: this branch should be used to build tokenizers with string tensor support (and without pack/unpack strings), necessary macros in this PR will activate element::string part in your PR

This branch can be built but I don't think this is still functional because too many changes happened in master and I haven't validated my recent merge. Will provide some coverage on that.

### CODE REVIEW

File: src/core/src/model.cpp [hunk 1]
```diff
@@ -443,6 +443,14 @@ void ov::Model::replace_parameter(size_t parameter_index, const shared_ptr<ov::o
...
 }
 
+void ov::Model::set_parameter(size_t parameter_index, const shared_ptr<ngraph::op::Parameter>& parameter) {
+    m_parameters[parameter_index] = parameter;
+}
+
+void ov::Model::set_result(size_t index, const shared_ptr<ngraph::op::Result>& result) {
+    m_results[index] = result;
+}
+
```
Comment (slyalin) [inline src/core/src/model.cpp:453]: @itikhono, consider adding these methods. There is no good alternative in the current API.

File: src/inference/include/openvino/runtime/infer_request.hpp [hunk 2]
```diff
@@ -32,6 +32,11 @@ class OPENVINO_RUNTIME_API InferRequest {
     std::shared_ptr<ov::IAsyncInferRequest> _impl;
     std::shared_ptr<void> _so;
 
+    // TODO: Should be moved deeply into plugins depending on their capabilities to handle arbitrary objects
+    // Map from name of a tensor to its value exposed to the user side
+    std::shared_ptr<std::map<std::string, Any>> _wrapped_objects =
+        std::make_shared<std::map<std::string, Any>>();
+
```
Comment (slyalin) [inline src/inference/include/openvino/runtime/infer_request.hpp:39]: Part of the CPU hack.

File: src/inference/src/dev/core_impl.cpp [hunk 3]
```diff
@@ -801,14 +801,78 @@ ov::SoPtr<ov::ICompiledModel> ov::CoreImpl::compile_model_with_preprocess(ov::Pl
...
+                // Store shape as a RT attribute, otherwise the validation of next nodes cannot deduce shape from a new parameter
+                PartialShape original_shape = result->get_output_partial_shape(0);
+                //parameter->get_rt_info()["original_partial_shape"] = original_shape;  // Not universal in case if the wrapping happens not for Parameter
+                result->set_output_type(0, element::u8, PartialShape{sizeof(void*)});   // FIXME: Weak, will be elimnated by the next validate and infer
+
+                // Add a new tensor name to recognize this changed parameter in the infer_request
+                mark_tensor(result->get_output_tensor(0), original_shape);
+            }
+        }
+
```
Comment (slyalin) [inline src/inference/src/dev/core_impl.cpp:875]: Part of the CPU hack.

File: src/inference/src/infer_request.cpp [hunk 4]
```diff
@@ -56,6 +56,16 @@ inline bool getPort(ov::Output<const ov::Node>& res_port,
...
+// TODO: Extend to any objects, not only strings
+bool is_object_wrapper(const ov::Output<const ov::Node>& port) {
+    const auto& names = port.get_names();
+    auto it = std::find_if(
+        names.begin(),
+        names.end(),
+        [](const std::string& name) {return name.find("__overriden_string_port_prefix__") == 0; });
+    return names.end() != it;
+}
+
```
Comment (slyalin) [inline src/inference/src/infer_request.cpp:68]: Part of the CPU hack.

File: src/inference/src/infer_request.cpp [hunk 5]
```diff
@@ -71,7 +81,18 @@ InferRequest::InferRequest(const std::shared_ptr<ov::IAsyncInferRequest>& impl,
...
+            std::cerr << "Hacking string tensor in the infer_request, tensor name: " << port.get_any_name() << "\n";
+            // Assume get_any_name is deterministic within one process
+            auto& ref = (*_wrapped_objects)[port.get_any_name()];
+            ref = tensor;
+            auto string_tensor = &ref.as<Tensor>();
+            *reinterpret_cast<Tensor**>(_impl->get_tensor(port)->data<uint8_t>()) = string_tensor;
+        } else {
+            _impl->set_tensor(port, get_tensor_impl(tensor));
+        }
+    });
```
Comment (slyalin) [inline src/inference/src/infer_request.cpp:95]: Part of the CPU hack.

File: src/inference/src/infer_request.cpp [hunk 6]
```diff
@@ -184,6 +205,20 @@ Tensor InferRequest::get_tensor(const ov::Output<const ov::Node>& port) {
...
+            // Assume get_any_name is deterministic within one process
+            if(_wrapped_objects->count(port.get_any_name()) == 0) {
+                // TODO: Set shape depending on the original model
+                (*_wrapped_objects)[port.get_any_name()] = Tensor(element::string, Shape{0});
+            }
+            auto string_tensor = &(*_wrapped_objects)[port.get_any_name()].as<Tensor>();
+            *reinterpret_cast<Tensor**>(tensor->data<uint8_t>()) = string_tensor;
+            std::cerr << string_tensor << "\n";
+            return *string_tensor;   // TODO: _so initialization?
+        }
```
Comment (slyalin) [inline src/inference/src/infer_request.cpp:221]: Part of the CPU hack.

File: src/plugins/intel_cpu/src/dnnl_extension_utils.cpp [hunk 7]
```diff
@@ -98,7 +99,7 @@ InferenceEngine::Precision DnnlExtensionUtils::DataTypeToIEPrecision(memory::dat
         case memory::data_type::u4:
             return InferenceEngine::Precision::U4;
         case memory::data_type::undef:
-            return InferenceEngine::Precision::UNSPECIFIED;
+            return InferenceEngine::Precision::UNSPECIFIED; // TODO: How about reverse conversion of Precision::STRING?
```
Comment (slyalin) [inline src/plugins/intel_cpu/src/dnnl_extension_utils.cpp:102]: Probably part of the CPU hack -- don't know how are you going to represent string tensor inside.

File: samples/cpp/string_support_test.cpp [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+        auto pitensor = &inputs[0];
+
+        if(pitensor->get_element_type() == ov::element::u8 && pitensor->get_byte_size() == sizeof(void*)) {
+            std::cerr << "[ DEBUG ] StringTensorOp::evaluate detected CPU hack at input, unwrapping\n";
+            auto data = *reinterpret_cast<const void* const*>(pitensor->data());
+            if(data != nullptr) {
+                pitensor = reinterpret_cast<const ov::Tensor*>(data);
+            }
+        }
```
Comment (slyalin) [inline samples/cpp/string_support_test.cpp:58]: Part of the CPU hack.

File: samples/cpp/string_support_test.cpp [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    ov::Core core;
+    auto compiled = core.compile_model(model, "CPU");
+
+    std::cout << "Compiled model input/output type: " << compiled.input().get_element_type() << "/" << compiled.output().get_element_type() << "\n";
+
+    std::cout << "--------\n";
+    auto request = compiled.create_infer_request();
+
+    request.set_input_tensor(init_tensor()); // required for CPU hack, otherwise output tensor won't be correctly allocated
+    auto output_tensor = request.get_output_tensor();   // trigger wrapping logic before infer request
```
Comment (slyalin) [inline samples/cpp/string_support_test.cpp:147]:
```suggestion
    request.set_input_tensor(init_tensor());
    auto output_tensor = request.get_output_tensor();   // trigger wrapping logic before infer request
```

File: samples/cpp/string_support_test.cpp [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    ov::Core core;
+    auto compiled = core.compile_model(model, "CPU");
+
+    std::cout << "Compiled model input/output type: " << compiled.input().get_element_type() << "/" << compiled.output().get_element_type() << "\n";
+
+    std::cout << "--------\n";
+    auto request = compiled.create_infer_request();
+
+    request.set_input_tensor(init_tensor());
+    auto output_tensor = request.get_output_tensor();   // trigger wrapping logic before infer request
```
Comment (slyalin) [inline samples/cpp/string_support_test.cpp:147]: Part of the CPU hack.

File: samples/cpp/string_support_test.cpp [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            auto data = *reinterpret_cast<void* const*>(potensor->data());
+            if(data != nullptr) {
+                potensor = reinterpret_cast<ov::Tensor*>(data);
+            }
+        }
+
+        auto output_tensor = *potensor;
+
+        std::cerr << "Output tensor type after optional unwrapping: " << output_tensor.get_element_type() << " and shape " << output_tensor.get_shape() << "\n";
+        output_tensor.set_shape(input_tensor.get_shape());
```
Comment (slyalin) [inline samples/cpp/string_support_test.cpp:78]: Part of the CPU hack: shouldn't be required as validate_infer_types will bring correctly computed shape and I believe CPU plugin should provide correctly shaped tensors for outputs.

File: samples/cpp/string_support_test.cpp [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+    void validate_and_infer_types() override {
+        // The conditions below is a part of CPU hack to support a wrapped string tensor as a u8 tensor with a pointer
+        if(get_input_size() == 1 && get_input_element_type(0) == ov::element::u8) {
+            std::cerr << "[ DEBUG ] StringTensorOp::validate_and_infer_types detected CPU hack at input, still providing element::string in the output\n";
+        }
+        if(get_output_size() == 1 && get_output_element_type(0) == ov::element::u8) {
+            std::cerr << "[ DEBUG ] StringTensorOp::validate_and_infer_types detected CPU hack at output, not overriding output type\n";
+            return;
+        }
```
Comment (slyalin) [inline samples/cpp/string_support_test.cpp:29]: Part of the CPU hack.

**Comment by slyalin** [comment]
In the recent commits, numpy.bytes_ arrays have been supported as ov.Tensor initializer and ov.Tensor.data accessor. As we are using numpy.bytes_ which is more compact representation than numpy.str_, we need to call decode/encode explicitly if data is required to be represented as normal string on the user side. Example:

```
import numpy as np
import torch
import openvino as ov

# Build a simple Parameter(string) -> Result(string) model with the help of torch.jit
ov_model = ov.convert_model(torch.jit.trace(lambda x: x, torch.Tensor(0)), input=ov.Type.string)
compiled = ov.compile_model(ov_model, 'CPU')

r = compiled(np.char.encode(np.array(['text', 'больше текста', 'jeszcze więcej słów'])))[0]
print(np.char.decode(r))
```

**Comment by rkazants** [comment]
@slyalin, do we need to update C binding with new type?
Also what do we expect pointed by `host_ptr` in ov::Tensor constructor `Tensor::Tensor(const element::Type& element_type, const Shape& shape, void* host_ptr, const Strides& byte_strides)`: already allocated std::string objects or raw data?

**Comment by slyalin** [comment]
> @slyalin, do we need to update C binding with new type? Also what do we expect pointed by `host_ptr` in ov::Tensor constructor `Tensor::Tensor(const element::Type& element_type, const Shape& shape, void* host_ptr, const Strides& byte_strides)`: already allocated std::string objects or raw data?

Already discussed offline, here is a summary: `host_ptr` points to already pre-initialized std::string objects array.

**Pull request closed without merging**